drv/fpga-server/src/main.rs

```diff
@@ -0,0 +1,272 @@
+
+#![no_std]
+#![no_main]
+
+use ringbuf::*;
+use userlib::*;
+use zerocopy::{byteorder, AsBytes, Unaligned, U16};
+
+use drv_fpga_api::*;
+use drv_fpga_devices::{ecp5, ecp5::Ecp5, ecp5_spi::Ecp5UsingSpi, Fpga};
+use drv_spi_api::{self as spi_api, Spi, SpiDevice};
+use drv_stm32xx_sys_api::{self as sys_api, Sys};
+use idol_runtime::{ClientError, Leased, LenLimit, R, W};
+
+task_slot!(SYS, sys);
+task_slot!(SPI, spi_driver);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum Trace {
+    None,
+    DeviceId(u32),
+    StartBitstreamLoad(BitstreamType),
+    ContinueBitstreamLoad(usize),
+    FinishBitstreamLoad,
+}
+ringbuf!(Trace, 16, Trace::None);
+
+#[export_name = "main"]
+fn main() -> ! {
+    let sys = Sys::from(SYS.get_task_id());
+    let spi = Spi::from(SPI.get_task_id()).device(0);
+
+    cfg_if::cfg_if! {
+        if #[cfg(target_board = "sidecar-1")] {
+            let driver = Ecp5UsingSpi {
+                sys,
+                spi,
+                done: sys_api::Port::J.pin(15),
+                init_n: sys_api::Port::J.pin(12),
+                program_n: sys_api::Port::J.pin(13),
+                design_reset_n: sys_api::Port::J.pin(14),
+            };
+            driver.configure_gpio();
+            let ecp5 = Ecp5::from(driver);
+        } else if #[cfg(target_board = "gimletlet-2")] {
+            let driver = Ecp5UsingSpi {
+                sys,
+                spi,
+                done: sys_api::Port::E.pin(15),
+                init_n: sys_api::Port::D.pin(12),
+                program_n: sys_api::Port::B.pin(10),
+                design_reset_n: sys_api::Port::D.pin(11),
+            };
+            driver.configure_gpio();
+            let ecp5 = Ecp5::from(driver);
+        } else {
+            compile_error!("Board is not supported by the task/fpga");
+        }
+    }
+
+    let mut incoming = [0u8; idl::INCOMING_SIZE];
+    let mut server = ServerImpl {
+        device: ecp5,
+        device_reset_ticks: ecp5::DEVICE_RESET_DURATION,
+        application: Spi::from(SPI.get_task_id()).device(1),
+        application_reset_ticks: ecp5::APPLICATION_RESET_DURATION,
+        buffer: [0u8; 128],
+        decompressor: None,
+    };
+
+    if let Ok(DeviceState::AwaitingBitstream) = server.device.device_state() {
+        ringbuf_entry!(Trace::DeviceId(server.device.device_id().unwrap()));
+    }
+
+    loop {
+        idol_runtime::dispatch(&mut incoming, &mut server);
+    }
+}
+
+struct ServerImpl<FpgaT: Fpga> {
+    device: FpgaT,
+    device_reset_ticks: u64,
+    application: SpiDevice,
+    application_reset_ticks: u64,
+    buffer: [u8; 128],
+    decompressor: Option<gnarle::Decompressor>,
+}
+
+type RequestError = idol_runtime::RequestError<FpgaError>;
+type ReadDataLease = LenLimit<Leased<R, [u8]>, 128>;
+type WriteDataLease = LenLimit<Leased<W, [u8]>, 128>;
+
+impl<FpgaT: Fpga> idl::InOrderFpgaImpl for ServerImpl<FpgaT> {
+    fn device_enabled(&mut self, _: &RecvMessage) -> Result<u8, RequestError> {
```

Comment:
I commented on the Idol file as well, but we should use `bool` instead of `u8` for the `*_enabled` / `set_*_enable` functions.

(and if that doesn't work, let me know, since that would be an Idolatry bug!)

Reply:
No, that seems to work as intended and makes this whole interface less awkward. Thanks!